SerenityOS/serenity
# New SystemServer, socket takeover and lazy spawning

Here's a little teaser:

```
Browser(19): CLocalSocket{0x015da112} connecting to /tmp/portal/protocol...
SystemServer(4): Ready to read on behalf of ProtocolServer
SystemServer(4): Spawning ProtocolServer
Process: New process PID=20 with name=SystemServer
Thread{0x00434da8}: New thread TID=0 in SystemServer(20)
SystemServer(20) do_exec(/bin/ProtocolServer): thread_count() = 1
VFS::create: 'rpc.20' in 4:1
ProtocolServer(20): Taking the socket over from SystemServer
sys$fcntl: fd=3, cmd=2, arg=1
Browser(19): CLocalSocket{0x015da112} connected ok!
```

here's another one:

![image](https://user-images.githubusercontent.com/10091584/69655289-4d326880-1087-11ea-9cb3-d697de6a4643.png)

This is the new SystemServer, inspired by systemd and launchd.

* Services are configured in `/etc/SystemServer.ini`
* It uses LibCore / `CObject`, it's introspectable; in the future it will be possible to make it accept commands (spawn this service, kill this service) and to surface this all in SystemMonitor
* It supports "socket takeover" (what systemd calls socket activation), where it creates sockets for servers itself, before starting anything else, in a clean and race-free manner, and then passes open fds to those sockets to the servers
* Going further with this idea, it can listen on those sockets and only spawn a server once some client tries to connect to its socket
* It watches its children and not only reaps them, but can also restart them

Side effects:
* "Portals" have been moved from `/tmp/fooportal` to `/tmp/portal/foo`
* LookupServer has been mostly ported to LibCore and cleaned up a little

To test:
* Kill the taskbar; it should respawn immediately
* Close the terminal, it should just go away without respawning
* Start the browser; this immediately starts the ProtocolServer
* Click on an external URL, it spawns LookupServer
* Kill the LookupServer, update the page
* `ins 4`
